library/core/src/ptr/non_null.rs

```diff
@@ -1795,12 +1832,75 @@ mod verify {
         }
     }
 
-    // pub const unsafe fn new(ptr: *mut T) -> Option<Self>
+    // pub const fn new(ptr: *mut T) -> Option<Self>
     #[kani::proof_for_contract(NonNull::new)]
     pub fn non_null_check_new() {
         let mut x: i32 = kani::any();
         let xptr = &mut x;
         let maybe_null_ptr =  if kani::any() { xptr as *mut i32 } else { null_mut() };
         let _ = NonNull::new(maybe_null_ptr);
     }
+
+    // pub const unsafe fn add(self, count: usize) -> Self
+    #[kani::proof_for_contract(NonNull::add)]
+    pub fn non_null_check_add() {
+        const SIZE: usize = 100000;
+        // Randomize pointer offset within array bound
+        let offset = kani::any_where(|x| * x <= SIZE as isize);
+        kani::assume(offset >= 0);
+        // Create a non-deterministic array of size SIZE
+        let arr: [i8; SIZE] = kani::any();  
+        // Get a raw pointer to the array
+        let raw_ptr: *mut i8 = arr.as_ptr() as *mut i8;  
+        // NonNull pointer to the random offset
+        let ptr = unsafe { NonNull::new(raw_ptr.offset(offset)).unwrap()};
+        // Create a non-deterministic count value
+        let count: usize = kani::any();  
+        
+        unsafe {
+            // Add a positive offset to pointer
+            let result = ptr.add(count);
+        }
+    }
+
+    // pub fn addr(self) -> NonZero<usize>
+    #[kani::proof_for_contract(NonNull::addr)]
+    pub fn non_null_check_addr() {  
+        // Create NonNull pointer & get pointer address
+        let mut x: i32 = kani::any();
```

Comment:
It's more general to do:
```rust
        let x= kani::any::<usize>() as *mut i32;
```
and pass the result to `NonNull::new()`.

Reply:
This works except we need to add `kani::assume(!x.is_null())` after this line so that we don't get failed check for calling `unwrap()` on a None object.

Reply:
Oh, yes. Adding the assumption is necessary.